### PR TITLE
Enable extra credentials for the system connector.

### PR DESCRIPTION
**Pull request by SthuthiGhosh9400**
## Description
It was identified that extra credentials were not being passed into the connector session when running queries against system.jdbc.tables. 
**fix:**
Modified the constructor call for the Identity class to pass extra credentials.

A use case is that `system.jdbc.tables` should not fail for any connectors (e.g., the Arrow Flight connector)

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by SthuthiGhosh9400** [comment]
@tdcmeehan 
Could you review this?

**Comment by tdcmeehan** [comment]
What fails without the extra credentials?  Can you put it in a unit test?

**Comment by tdcmeehan** [comment]
Please fix commit message according to the contributing guide.  Perhaps `Pass extra credentials to system connector`

**Comment by SthuthiGhosh9400** [comment]
> Please fix commit message according to the contributing guide. Perhaps `Pass extra credentials to system connector`


I have modified the commit message to `Pass extra credentials to system connector`
Thanks

**Review by ZacBlanco — APPROVED** [review]

**Review by ZacBlanco — CHANGES_REQUESTED** [review]
Can you add a unit test please?

**Comment by SthuthiGhosh9400** [comment]
@ZacBlanco @tdcmeehan 
I added unit test.
Could you please review ?

File: presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                extraCredentials.put("key1", "value1");
+                extraCredentials.put("key2", "value2");
+                return extraCredentials;
+            }
+        };
+
+        String validQueryId = "query_id_1";
+        QueryId queryId = new QueryId(validQueryId);
+
+        ConnectorSession session = new ConnectorSession() {
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java:70]: Please just use a TestingConnectorSession.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Please read the contributing guide, the commits need to be cleaned up.

File: presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java (hunk 1 above)
Comment (SthuthiGhosh9400) [reply presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java]: @tdcmeehan Used TestingConnectorSession. Could you have  a look?
Thanks.

File: presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        Optional<SelectedRole> selectedRole = Optional.empty();
+
+        ConnectorIdentity connectorIdentity = new ConnectorIdentity("user", principal, selectedRole) {
+            @Override
+            public Map<String, String> getExtraCredentials()
+            {
+                Map<String, String> extraCredentials = new HashMap<>();
+                extraCredentials.put("key1", "value1");
+                extraCredentials.put("key2", "value2");
+                return extraCredentials;
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java:56]:
```suggestion
                return ImmutableMap.of("key1", "value1", "key2", "value2");
```

File: presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        TransactionId transactionId = new TransactionId(UUID.randomUUID());
+
+        ConnectorTransactionHandle transactionHandle = new GlobalSystemTransactionHandle("connector-id", transactionId);
+
+        Optional<Principal> principal = Optional.of(() -> "principal");
+
+        Optional<SelectedRole> selectedRole = Optional.empty();
+
+        ConnectorIdentity connectorIdentity = new ConnectorIdentity("user", principal, selectedRole) {
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java:49]: Please use the full constructor and don't override extra credentials.

File: presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                        Object::toString,
+                        value -> value)));
+
+        Session resultSession = SystemConnectorSessionUtil.toSession(transactionHandle, testingConnectorSession);
+
+        assertNotNull(resultSession, "Session should not be null");
+        assertEquals(resultSession.getIdentity().getUser(), "user", "User should match");
+        assertEquals(resultSession.getIdentity().getPrincipal().get().getName(), "principal", "Principal should match");
+        assertEquals(resultSession.getIdentity().getExtraCredentials().get("key1"), "value1", "Extra credential for key1 should match");
+        assertEquals(resultSession.getIdentity().getExtraCredentials().get("key2"), "value2", "Extra credential for key2 should match");
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java:76]: Just remove these messages, they are not helpful because they don't tell you the mismatched values.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

**Comment by SthuthiGhosh9400** [comment]
@tdcmeehan 
I have addressed the comments. Kindly review them.

File: presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        Optional<SelectedRole> selectedRole = Optional.empty();
+
+        Map<String, TokenAuthenticator> extraAuthenticators = new HashMap<>();
+        extraAuthenticators.put("token1", new TokenAuthenticator() {
+            @Override
+            public Slice getAuthenticationToken(Slice authenticationTokenRequest)
+            {
+                return null;
+            }
+        });
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java:59]: Why is this needed?

Comment (SthuthiGhosh9400) [reply presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java]: Yeah, removed it since it’s not contributing to the behavior being tested.

File: presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                        Object::toString,
+                        value -> value)));
+
+        Session resultSession = SystemConnectorSessionUtil.toSession(transactionHandle, testingConnectorSession);
+
+        assertNotNull(resultSession);
+        assertEquals(resultSession.getIdentity().getUser(), "user");
+        assertEquals(resultSession.getIdentity().getPrincipal().get().getName(), "principal");
+        assertEquals(resultSession.getIdentity().getExtraCredentials().get("key1"), "value1");
+        assertEquals(resultSession.getIdentity().getExtraCredentials().get("key2"), "value2");
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/connector/system/TestSystemSession.java:73]:
```suggestion
        Map<String, String> extraCredentials = ImmutableMap.of("key1", "value1", "key2", "value2");
        String user = "user";
        String principal = "principal";

        ConnectorTransactionHandle transactionHandle = new GlobalSystemTransactionHandle("connector-id", new TransactionId(UUID.randomUUID()));

        ConnectorIdentity connectorIdentity = new ConnectorIdentity(
                user,
                Optional.of(() -> principal),
                Optional.empty(),
                extraCredentials,
                ImmutableMap.of(),
                Optional.of("selectedUserName"),
                Optional.of("reasonForSelection"));

        TestingConnectorSession testingConnectorSession = new TestingConnectorSession(connectorIdentity,
                ImmutableList.of(new PropertyMetadata<>("property-name",
                        "This is a description",
                        VarcharType.createUnboundedVarcharType(),
                        String.class,
                        "defaultValue",
                        false,
                        Object::toString,
                        value -> value)));

        Session resultSession = SystemConnectorSessionUtil.toSession(transactionHandle, testingConnectorSession);

        assertNotNull(resultSession);
        assertEquals(resultSession.getIdentity().getUser(), user);
        assertEquals(resultSession.getIdentity().getPrincipal().get().getName(), principal);
        assertEquals(resultSession.getIdentity().getExtraCredentials(), extraCredentials);
```

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

**Comment by SthuthiGhosh9400** [comment]
@tdcmeehan 
Modified the code in the TestSystemSession.java according to the suggestion. 
Thanks.

**Review by tdcmeehan — APPROVED** [review]

**Comment by tdcmeehan** [comment]
@ZacBlanco WDYT?

File: presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorSessionUtil.java [hunk 7]
```diff
@@ -34,7 +34,7 @@ public static Session toSession(ConnectorTransactionHandle transactionHandle, Co
     {
         TransactionId transactionId = ((GlobalSystemTransactionHandle) transactionHandle).getTransactionId();
         ConnectorIdentity connectorIdentity = session.getIdentity();
-        Identity identity = new Identity(connectorIdentity.getUser(), connectorIdentity.getPrincipal());
+        Identity identity = new Identity(connectorIdentity.getUser(), connectorIdentity.getPrincipal(), connectorIdentity.getExtraCredentials());
```
Comment (ZacBlanco) [inline presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorSessionUtil.java:37]: I'm not very familiar with this code, but I'm wondering if we should just populate the entire `Identity` object? There are fields within `Identity` that are the same as in `ConnectorIdentity` such as `extraAuthenticators`, `selectedUser`, and `reasonForSelect`. Should we also populate those too or would there be adverse side effects?

Comment (SthuthiGhosh9400) [reply presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorSessionUtil.java:37]: In this case, populating only the necessary fields in the Identity object should be sufficient. Since my goal is to ensure that the system connector (e.g., system.jdbc.tables) works across connectors like the Arrow Flight connector without failures, I focused on fields that are required for this functionality to work correctly.

Comment (SthuthiGhosh9400) [reply presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorSessionUtil.java:37]: I was thinking like, If the system connector doesn't explicitly require fields like extraAuthenticators, selectedUser, or reasonForSelect, there is no need to pass them. These fields could typically be used in more specific cases, such as when additional authentication mechanisms are involved, or when user impersonation is necessary.

extraAuthenticators: This is used when additional authentication layers are needed. If the system connector or your use case doesn’t involve extra authentication, passing this field is unnecessary.

selectedUser: This is mainly relevant for scenarios where user impersonation or proxy users are involved. If your system connector doesn’t handle such cases, it’s safe to omit this field.

reasonForSelect: This field provides context for why a user switch occurred.

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorSessionUtil.java:37]: Thanks for the context. I think it's fine to leave the additional fields out for now

Comment (SthuthiGhosh9400) [reply presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorSessionUtil.java:37]: When you have a chance, could you approve the changes if everything looks good? @ZacBlanco

**Review by ZacBlanco — APPROVED** [review]
